packages/ckb-sdk-core/src/signWitnesses.ts

```diff
@@ -2,52 +2,65 @@ import { ArgumentRequired } from '@nervosnetwork/ckb-sdk-utils/lib/exceptions'
 import signWitnessGroup from './signWitnessGroup'
 import groupScripts from './groupScripts'
 
-type Key = string
+type SignatureProvider = string | ((message: string | Uint8Array) => string)
 type LockHash = string
 type TransactionHash = string
 type CachedLock = Pick<CachedCell, 'lock'>
 
 export interface SignWitnesses {
-  (key: Key): (params: { transactionHash: TransactionHash; witnesses: StructuredWitness[] }) => StructuredWitness[]
-  (key: Map<LockHash, Key>): (params: {
+  (key: SignatureProvider): (params: {
+    transactionHash: TransactionHash
+    witnesses: StructuredWitness[]
+  }) => StructuredWitness[]
+  (key: Map<LockHash, SignatureProvider>): (params: {
     transactionHash: TransactionHash
     witnesses: StructuredWitness[]
     inputCells: CachedLock[]
+    skipMissingKeys: boolean
   }) => StructuredWitness[]
-  (key: Key | Map<LockHash, Key>): (params: {
+  (key: SignatureProvider | Map<LockHash, SignatureProvider>): (params: {
     transactionHash: TransactionHash
     witnesses: StructuredWitness[]
     inputCells?: CachedLock[]
+    skipMissingKeys?: boolean
   }) => StructuredWitness[]
 }
 
 export const isMap = <K = any, V = any>(val: any): val is Map<K, V> => {
   return val.size !== undefined
 }
 
-const signWitnesses: SignWitnesses = (key: Key | Map<LockHash, Key>) => ({
+const signWitnesses: SignWitnesses = (key: SignatureProvider | Map<LockHash, SignatureProvider>) => ({
   transactionHash,
   witnesses = [],
   inputCells = [],
+  skipMissingKeys = false,
 }: {
   transactionHash: string
   witnesses: StructuredWitness[]
   inputCells: CachedLock[]
+  skipMissingKeys: boolean
 }) => {
-  if (!key) throw new ArgumentRequired('Private key')
+  if (!key) throw new ArgumentRequired('Signature provider')
   if (!transactionHash) throw new ArgumentRequired('Transaction hash')
   if (!witnesses.length) throw new Error('Witnesses is empty')
 
   if (isMap(key)) {
     const rawWitnesses = witnesses
     const restWitnesses = witnesses.slice(inputCells.length)
     const groupedScripts = groupScripts(inputCells)
+
     groupedScripts.forEach((indices, lockhash) => {
       const sk = key.get(lockhash)
       if (!sk) {
```

Comment:
The nested `if` could be simplified as `if (!sk && !skipMissingKeys)`

Reply:
If `sk` is not presented it should either throw an error or return the current execution. Otherwise, proceed and do the signing.

Reply:
Sure, it does need to return